// 1.0.3/cloudwise.js
'use strict'

var http = require('http');
var rf = require("fs");
var logger = require('./lib/logger').child({ component: 'cloudwise' });
var hostkey = ""
var hostid = ""
function trim(str) { //删除左右两端的空格
    return str.replace(/(^s*)|(s*$)/g, "");
}

global.isMonitor = true;
global.lock = 0;
global.confPath = JSON.parse(rf.readFileSync("./node_modules/NodejsAgent/SmartAgentPath.txt", "utf-8"))["appconf"]
function monitor() {
    rf.readFile(global.confPath, function (err, data) {
        if (err) return logger.error(err);
        if (data.length > 0) {
            var obj = JSON.parse(data)
            global.isMonitor = obj["isopen"];
            global.TransferProtocol = obj["TransferProtocol"]
            global.SendProxyHost = obj["SendProxyHost"]
            global.SendProxyUdpPort = obj["SendProxyUdpPort"]
            global.SendProxyTcpPort = obj["SendProxyTcpPort"]
            global.HostKey = obj["HostKey"]
            global.HostId = obj["HostId"]
            global.lock = 0;
        }
        else {
            logger.info("try again to read the appconf change ")
            global.lock += 1;
            if (global.lock > 5) {
                logger.info("change appconf fail cause have try " + global.lock + " times but still not get values");
                return;
            }
            else {
                monitor()
            }
        }
    });
}
monitor();
rf.watch(global.confPath, function (event, filename) {
    logger.info('event is: ' + event); if (filename) {
        monitor()
        logger.info('filename provided: ' + filename);
    } else {
        logger.info('filename not provided');
    }
});


function cloudwiseConf() {
    this.gethostid = function () {
        var options = {
            hostname: '127.0.0.1',
            port: 27789,
            path: '/smartagent/getHostId',
            method: 'GET'
        };




        var req1 = http.request(options, function (res1) {
            res1.setEncoding('utf8');
            res1.on('data', function (chunk) {
                hostid = chunk
                // console.log("hostkey=" + hostkey);
            });
        });

        req1.end();
        return hostid
    }

    this.gethostkey = function () {
        var options2 = {
            hostname: '127.0.0.1',
            port: 27789,
            path: '/smartagent/getHostKey',
            method: 'GET'
        };

        var req2 = http.request(options2, function (res2) {
            res2.setEncoding('utf8');
            res2.on('data', function (chunk2) {
                hostkey = chunk2
                //console.log("hostid=" + hostid);
            });
        });

        req2.end();
        return hostkey
    }

}

module.exports = cloudwiseConf;
