// 1.0.3/lib/metrics/iodata.js
'use strict';

var RequestMethod = require('../collector/request-method.js');
var cloudconf = require('../../cloudwise.js');
var sources = require('./source.js');


var remote = new RequestMethod({
    method: 'POST',
    host: '127.0.0.1',
    port: '26789',
    path: '/send',
    ssl: false
});


/**
 * 存储http请求数据
 * @constructor
 */
function iodata() {

}




/**
 *发送http记录数据
 */
iodata.prototype.send = function (segment) {
	var source=new sources();
	
    source.setHost_id(cloudconf.cloudwiseConf.HostId);
    source.setAppName(cloudconf.cloudwiseConf.app_name);
    source.setIp(getIPAdress());
    source.setNodeTopic(cloudconf.cloudwiseConf.HostkKey + '_1005_' + (new Date).getTime());
    source.addMaps(segment.name, segment.timer.duration, "", "", "", "", "", "", segment.souceCode, 0, "", "", JSON.stringify(segment.args[0]), JSON.stringify(segment.args[1]), "1", segment.timer.start, segment.host + ":" + segment.port, "205");
    source.addResources('2301', segment.host, segment.dbName, segment.port, segment.userName);
    remote.invoke(source.getWowMap(), function (res) {
        var body = '';
        if (res.statusCode == 200) {
            res.on('data', function (data) {
                body += data;
            }).on('end', function () {
                console.log(body);
            })
        } else {
            console.log('request is error!')
        }
    });
};

function getIPAdress() {
    var interfaces = require('os').networkInterfaces();
    for (var devName in interfaces) {
        var iface = interfaces[devName];
        for (var i = 0; i < iface.length; i++) {
            var alias = iface[i];
            if (alias.family === 'IPv4' && alias.address !== '127.0.0.1' && !alias.internal) {
                return alias.address;
            }
        }
    }
}

module.exports = iodata;
